packages/react-native/ReactCommon/react/nativemodule/idlecallbacks/NativeIdleCallbacks.cpp

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <react/renderer/runtimescheduler/RuntimeScheduler.h>
+#include <react/renderer/runtimescheduler/RuntimeSchedulerBinding.h>
+#include "NativeIdleCallbacks.h"
+
+#ifdef RN_DISABLE_OSS_PLUGIN_HEADER
+#include "Plugins.h"
+#endif
+
+std::shared_ptr<facebook::react::TurboModule> NativeIdleCallbacksModuleProvider(
+    std::shared_ptr<facebook::react::CallInvoker> jsInvoker) {
+  return std::make_shared<facebook::react::NativeIdleCallbacks>(
+      std::move(jsInvoker));
+}
+
+namespace facebook::react {
+
+class IdleTaskRef : public jsi::NativeState {
+public:
+  IdleTaskRef(std::shared_ptr<Task> task): task(task) {}
+  virtual ~IdleTaskRef() noexcept override {}
+  
+  std::shared_ptr<Task> task;
+};
+
+jsi::Function makeTimeRemainingFunction(
+    jsi::Runtime& runtime,
+    std::shared_ptr<RuntimeScheduler> runtimeScheduler,
+    RuntimeSchedulerTimePoint deadline) {
+  return jsi::Function::createFromHostFunction(
+    runtime,
+    jsi::PropNameID::forAscii(runtime, "timeRemaining"),
+    0,
+    [&](
+      jsi::Runtime& runtime,
+      const jsi::Value& /* unused */,
+      const jsi::Value* /* unused */,
+      size_t /* unused */) noexcept -> jsi::Value {
+        if (runtimeScheduler->getShouldYield()) {
+          return jsi::Value(runtime, 0.0);
+        }
+              
+        auto now = runtimeScheduler->now();
+        double remainingTime = std::chrono::duration_cast<std::chrono::milliseconds>(deadline - now).count();
```

Comment:
For future reference, this `deadline` here is garbage because it's passed as a reference to this host function and the value is being cleared from the stack of `makeTimeRemainingFunction`, which made `timeRemaining()` always return 0. The fix is to copy the value to the lambda.

I'll fix this before landing.